core/core-configuration-layer.el

```diff
@@ -1866,7 +1870,18 @@ FILE-TO-LOAD is an explicit file to load after the installation."
             (1+ configuration-layer-error-count))
     (setq configuration-layer-error-count 1)))
 
+(defun configuration-layer/pin-package (package archive)
+  "Pin PACKAGE to ARCHIVE.
+
+When PACKAGE is pinned to specific ARCHIVE, it is installed
+exactly from that ARCHIVE.
+
+ARCHIVE must be one of the following symbols: melpa, org, gnu,
+melpa-stable."
+  (unless (memq archive '(melpa org gnu melpa-stable))
```

Comment:
Should use `package-archives` instead of hard-coded archive names:
```elisp
;; if archive is a string
(assoc archive package-archives)
;; if archive is a symbol
(assoc (symbol-name archive) package-archives)
```